src/lib/controller/message_handler.ts

```diff
@@ -118,3 +159,71 @@ export class ControllerMessageHandler {
     }
   }
 }
+
+function processInclusionOptions(
+  clientsController: ClientsController,
+  client: Client,
+  message:
+    | IncomingCommandControllerBeginInclusion
+    | IncomingCommandControllerBeginInclusionLegacy
+    | IncomingCommandControllerReplaceFailedNode
+    | IncomingCommandControllerReplaceFailedNodeLegacy
+): InclusionOptions | ReplaceNodeOptions {
+  // Schema 8+ inclusion handling
+  if ("options" in message) {
+    const options = message.options;
+    if (
+      options.strategy === InclusionStrategy.Default ||
+      options.strategy === InclusionStrategy.Security_S2
+    ) {
+      options.userCallbacks = {
+        grantSecurityClasses: (
+          requested: InclusionGrant
+        ): Promise<InclusionGrant | false> => {
+          clientsController.grantSecurityClassesPromise =
+            createDeferredPromise();
+          clientsController.grantSecurityClassesPromise.finally(() => {
+            delete clientsController.grantSecurityClassesPromise;
+          });
+          client.sendEvent({
+            source: "controller",
+            event: "grant security classes",
+            requested: requested as any,
+          });
+
+          return clientsController.grantSecurityClassesPromise;
+        },
+        validateDSKAndEnterPIN: (dsk: string): Promise<string | false> => {
+          clientsController.validateDSKAndEnterPinPromise =
+            createDeferredPromise();
+          clientsController.validateDSKAndEnterPinPromise.finally(() => {
+            delete clientsController.validateDSKAndEnterPinPromise;
+          });
+          client.sendEvent({
+            source: "controller",
+            event: "validate dsk and enter pin",
+            dsk,
+          });
+          return clientsController.validateDSKAndEnterPinPromise;
+        },
+        abort: (): void => {
+          clientsController.grantSecurityClassesPromise?.reject("aborted");
+          clientsController.validateDSKAndEnterPinPromise?.reject("aborted");
+          client.sendEvent({
+            source: "controller",
+            event: "inclusion aborted",
+          });
+        },
```

Comment:
Keep track of created promises in local scope to avoid rejecting promises created later.

```suggestion
      let grantSecurityClassesPromise: Promise<InclusionGrant | false> | undefined;
      let validateDSKAndEnterPinPromise: Promise<string | false> | undefined;

      options.userCallbacks = {
        grantSecurityClasses: (
          requested: InclusionGrant
        ): Promise<InclusionGrant | false> => {
          grantSecurityClassesPromise =
            createDeferredPromise();
          clientsController.grantSecurityClassesPromise =
            grantSecurityClassesPromise;
          grantSecurityClassesPromise.finally(() => {
            delete clientsController.grantSecurityClassesPromise;
          });
          client.sendEvent({
            source: "controller",
            event: "grant security classes",
            requested: requested as any,
          });

          return grantSecurityClassesPromise;
        },
        validateDSKAndEnterPIN: (dsk: string): Promise<string | false> => {
          validateDSKAndEnterPinPromise =
            createDeferredPromise();
          clientsController.validateDSKAndEnterPinPromise =
            validateDSKAndEnterPinPromise;
          validateDSKAndEnterPinPromise.finally(() => {
            delete clientsController.validateDSKAndEnterPinPromise;
          });
          client.sendEvent({
            source: "controller",
            event: "validate dsk and enter pin",
            dsk,
          });
          return validateDSKAndEnterPinPromise;
        },
        abort: (): void => {
          grantSecurityClassesPromise?.reject("aborted");
          validateDSKAndEnterPinPromise?.reject("aborted");
          client.sendEvent({
            source: "controller",
            event: "inclusion aborted",
          });
        },
```

Reply:
Wouldn't hurt to also add extra guard checks when `grantSecurityClasses` starts to ensure that it hasn't been called before and clean up is happening as expected:

```ts
if (grantSecurityClassesPromise !== undefined) {
  throw new Error("grantSecurityClassesPromise expected to be undefined");
}
if (client.grantSecurityClassesPromise !== undefined) {
  throw new Error("client.grantSecurityClassesPromise expected to be undefined");
}
```

Reply:
If I keep them locally scoped I can't guard calling `ControllerCommand.grantSecurityClasses` or `ControllerCommand.validateDSKAndEnterPIN` commands being called at the wrong time. Is that OK?

Reply:
You keep them locally scoped but also store them on `client`. That way you know which one you created and make sure you always only interact with your promises.

Reply:
ah I see what you did 👍🏾

Reply:
hmm running into some typing issues. Specifically on the reject: `Property 'reject' does not exist on type 'Promise<false | InclusionGrant>'.ts(2339)`

Given that only one inclusion process can be going at a time, should we just guard against beginInclusion being called while these promises on clientsController are not undefined?

Reply:
nvm, figured it out. I added a guard anyway